### ISSUE TITLE
Expired accessToken being used in authorization header while requesting for a new accessToken

### ISSUE THREAD

**Issue by amit-jha-cmd** [transistorsoft/flutter_background_geolocation#832]
## Your Environment
* Plugin version: 4.7.2
* Platform: iOS or Android
* OS version: All
* Device manufacturer / model: All
* Flutter info (`flutter doctor`):
```
[√] Flutter (Channel stable, 3.0.5, on Microsoft Windows [Version 10.0.22000.978], locale en-IN)
Checking Android licenses is taking an unexpectedly long time...[√] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[√] Chrome - develop for the web
[√] Android Studio (version 2021.2)
[√] IntelliJ IDEA Ultimate Edition (version 2022.1)
[√] VS Code (version 1.71.2)
[√] Connected device (4 available)
[√] HTTP Host Availability

```
* Plugin config:
```dart <-- Syntax highlighting: DO NOT REMOVE -->
        Config(
          desiredAccuracy: Config.DESIRED_ACCURACY_HIGH,
          locationUpdateInterval: WrenchConstants.locationUpdateInterval,
          url: '${Endpoint.addjobtrace}',
          method: 'POST',
          stationaryRadius: WrenchConstants.stationaryRadius,
          fastestLocationUpdateInterval:
              WrenchConstants.fastestLocationUpdateInterval,
          minimumActivityRecognitionConfidence:
              WrenchConstants.minimumActivityRecognitionConfidence,
          backgroundPermissionRationale: PermissionRationale(
            message: WrenchConstants.backgroundLocationPermission,
            title: WrenchConstants.backgroundLocationTitle,
          ),
          stopOnTerminate: false,
          startOnBoot: true,
          enableHeadless: true,
          headers: {
            'cognito_sso': true,
          },
          schedule: ['$startTime $endTime'],
          params: {
            'providerId': 1234,
          },
          authorization: Authorization(
            accessToken: awsCognitoUserPoolTokens?.accessToken,
            refreshToken: awsCognitoUserPoolTokens?.refreshToken,
            refreshUrl: LocationTrackingAuthConfig.getRefreshUrl(),
            refreshPayload: {
              "grant_type": "refresh_token",
              "client_id": _clientId,
              "client_secret": _clientSecret,
              "refresh_token": "{refreshToken}",
            },
          ),
        )
```

## Expected Behavior
The plugin requests for the new token when it encounters a 401 error (that means that the old token is expired). So, while making a POST request to `refreshUrl` the old accessToken should not be included in the `Authorization` header for 2 reasons:
1. The token is expired
2. Popular vendors like [amazon cognito](https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html) support `client_secret_basic` and ` client_secret_post` policies to get access token. They rely on the presence of the `Authorization` header to determine which policy should be used to authenticate the request

## Actual Behavior
While making a POST request to `refreshUrl` the `Authorization` header with the expired token is used which tricks cognito into using the wrong policy to authenticate the request which ultimately ends with a failure.

I'm trying to configure the plugin to not include the `Authorization` header while requesting a new access token but I'm unable to find a solution.

It would be good if you could help me with this issue.

**Comment by amit-jha-cmd** [comment]
i propose having support for `refreshHeader` like this:

```dart
Authorization(
  accessToken: awsCognitoUserPoolTokens?.accessToken,
  refreshToken: awsCognitoUserPoolTokens?.refreshToken,
  refreshUrl: LocationTrackingAuthConfig.getRefreshUrl(),
  refreshHeader: {
   'Authentication': "{accessToken}"
  }
  refreshPayload: {
    "grant_type": "refresh_token",
    "client_id": _clientId,
    "client_secret": _clientSecret,
    "refresh_token": "{refreshToken}",
  },
),
```

If we omit `refreshHeader` in the Authorization object then it should fall back to the current behavior otherwise it should use the user-defined template

**Comment by christocracy** [comment]
Do you have an api specification doc for this?

**Comment by amit-jha-cmd** [comment]
> Do you have an api specification doc for this?

cognito providers [/oauth2/token](https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html) endpoint to get a new access token using the refresh token, client_id, client_secret.

**Comment by christocracy** [comment]
So you’d call this new `Authorization.strategy: "aws”`?

**Comment by christocracy** [comment]
Or “cognito”?

**Comment by amit-jha-cmd** [comment]
no. The strategy will be JWT which is already supported. The problem is `Authorization` header is being sent to cognito which tricks it into using `client_secret_basic` authentication policy.

**Comment by christocracy** [comment]
And with this new “cognito” Authorization.strategy, you simply want the Authorization header excluded on refreshUrl requests.

> no. The strategy will be JWT

no, I think it will be a new strategy called “cognito”.  There’s no need to add extra complexity of a new Config option Authorization.refreshHeader

**Comment by amit-jha-cmd** [comment]
if I can somehow configure the plugin to not send the `Authorization` header while requesting a new token then my problem will be solved.

**Comment by amit-jha-cmd** [comment]
> And with this new “cognito” Authorization.strategy, you simply want the Authorization header excluded on refreshUrl requests.
> 
> > no. The strategy will be JWT
> 
> no, I think it will be a new strategy called “cognito”. There’s no need to add extra complexity of a new Config option Authorization.refreshHeader

sure. this works

**Comment by christocracy** [comment]
It will be ready in about two weeks.

**Comment by amit-jha-cmd** [comment]
Thanks for the prompt response!

**Comment by amit-jha-cmd** [comment]
Thank you!

**Comment by christocracy** [comment]
the default will be:  'Authorization': "{accessToken}`.  Typical users will not even define `refreshHeaders`.

For your use-case, you will provide `refreshHeaders` with an empty `{}`, supplying *no* headers.

```dart
Authorization(
  accessToken: awsCognitoUserPoolTokens?.accessToken,
  refreshToken: awsCognitoUserPoolTokens?.refreshToken,
  refreshUrl: LocationTrackingAuthConfig.getRefreshUrl(),
  refreshHeaders: {},  // <--- Provide NO headers.
  refreshPayload: {
    "grant_type": "refresh_token",
    "client_id": _clientId,
    "client_secret": _clientSecret,
    "refresh_token": "{refreshToken}",
  },
),
```

**Comment by amit-jha-cmd** [comment]
understood. sounds good

**Comment by christocracy** [comment]
Pushed this to branch `release-4.7.3`.  Please test by installing from Git:

*pubspec.yaml*
```yaml
dependencies:
  flutter_background_geolocation:
    git:
      url:  https://github.com/transistorsoft/flutter_background_geolocation.git
      ref:  release-4.7.3
```

**Comment by amit-jha-cmd** [comment]
I can confirm that this works as expected 👍🏾 

![image](https://user-images.githubusercontent.com/26526504/192366921-f3fa4b0e-fe63-4be0-b18e-acf872a76833.png)

**Comment by christocracy** [comment]
Super.  I will release to pub.dev tomorrow.

**Comment by amit-jha-cmd** [comment]
Thank you!

**Comment by christocracy** [comment]
You tested both iOS and Android?

**Comment by amit-jha-cmd** [comment]
No. I've only tested android. I'll do a quick round of testing on iOS and get back to you

**Comment by amit-jha-cmd** [comment]
iOS build failed with the following stack trace (GitHub won't let me paste it because it exceeds the character limit so attached it as a file)

-> [trace.txt](https://github.com/transistorsoft/flutter_background_geolocation/files/9649648/trace.txt) 

build fails even if I remove the `Authorization` object.

**Comment by amit-jha-cmd** [comment]
I executed this build remotely on a system sitting in my office. I can give you a more detailed stack trace tomorrow (from XCode), if required.

**Comment by christocracy** [comment]
That version was built with latest XCode.  That often causes problems unless you build your app with the same version that TSLocationManager.framework was built with.

**Comment by amit-jha-cmd** [comment]
In that case I'll upgrade XCode and retry the build tomorrow. I'll ping here once I'm able to build/test.

**Comment by christocracy** [comment]
published to `4.8.0`.  I decided to bump to `4.8.0` rather than use minor version `4.7.3` due to re-building with *XCode 14*.
